tests/integration/test_nginx_route.py

```diff
@@ -0,0 +1,192 @@
+
+
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_ingress_connectivity():
+    """
+    arrange: given charm has been built and deployed.
+    act: access ingress IP address with correct host name in HTTP headers.
+    assert: HTTP request should be forwarded to the application.
+    """
+    response = requests_get("http://127.0.0.1/ok", host_header="any")
+
+    assert response.text == "ok"
+    assert response.status_code == 200
+    assert (
+        requests_get("http://127.0.0.1/ok", host_header="any.other", retry_timeout=0).status_code
+        == 404
+    )
+
+
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_update_service_hostname(ops_test: OpsTest):
+    """
+    arrange: given charm has been built and deployed.
+    act: update the service-hostname option in any-charm.
+    assert: HTTP request with the service-hostname value as the host header should be forwarded
+        to the application correctly.
+    """
+    new_hostname = "any.other"
+    assert ops_test.model
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
+        {"src-overwrite": gen_src_overwrite(service_hostname=new_hostname)}
+    )
+    await ops_test.model.wait_for_idle(status="active")
+
+    response = requests_get("http://127.0.0.1/ok", host_header=new_hostname)
+    assert response.text == "ok"
+    assert response.status_code == 200
+
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
+        {"src-overwrite": gen_src_overwrite()}
+    )
+    await ops_test.model.wait_for_idle(status="active")
+
+
+@pytest.mark.usefixtures("build_and_deploy")
+async def test_update_additional_hosts(ops_test: OpsTest, run_action):
+    """
+    arrange: given charm has been built and deployed,
+    act: update the additional-hostnames option in the nginx-route relation using any-charm.
+    assert: HTTP request with the additional-hostnames value as the host header should be
+        forwarded to the application correctly. And the additional-hostnames should exist
+        in the nginx-route relation data.
+    """
+
+    async def get_relation_data():
+        action_result = await run_action(ANY_APP_NAME, "get-relation-data")
+        relation_data = json.loads(action_result["relation-data"])[0]
+        return relation_data["application_data"]["any"]
+
+    new_hostname = "any.new"
+    assert ops_test.model
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
+        {"src-overwrite": gen_src_overwrite(additional_hostnames=new_hostname)}
+    )
+
+    await ops_test.model.wait_for_idle(status="active")
+    response = requests_get("http://127.0.0.1/ok", host_header=new_hostname)
+    assert response.text == "ok"
+    assert response.status_code == 200
+    assert "additional-hostnames" in await get_relation_data()
+
+    await ops_test.model.applications[ANY_APP_NAME].set_config(
+        {"src-overwrite": gen_src_overwrite()}
+    )
+
+    await ops_test.model.wait_for_idle(status="active")
```

Comment:
Similar to above, should this be done in a fixture? It looks like cleanup after the test has run

Reply:
Same as the reply above.